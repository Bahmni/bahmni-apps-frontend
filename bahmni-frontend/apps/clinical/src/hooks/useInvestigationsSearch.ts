import { useEffect, useState, useMemo } from 'react';
import {
  useTranslation,
  getFormattedError,
  getFlattenedInvestigations,
  type FlattenedInvestigations
} from '@bahmni-frontend/bahmni-services';
import useDebounce from './useDebounce';

interface UseInvestigationsSearchResult {
  investigations: FlattenedInvestigations[];
  isLoading: boolean;
  error: Error | null;
}

const useInvestigationsSearch = (
  searchTerm: string = '',
): UseInvestigationsSearchResult => {
  const { t } = useTranslation();
  const [investigations, setInvestigations] = useState<
    FlattenedInvestigations[]
  >([]);
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState<Error | null>(null);
  const debouncedSearchTerm = useDebounce(searchTerm);
  useEffect(() => {
    const fetchInvestigations = async () => {
      try {
        setIsLoading(true);
        setError(null);
        const fetchedInvestigations = await getFlattenedInvestigations();
        setInvestigations(fetchedInvestigations);
      } catch (err) {
        const formattedError = getFormattedError(err);
        setError(new Error(formattedError.message));
      } finally {
        setIsLoading(false);
      }
    };
    fetchInvestigations();
  },[]);

  const filteredInvestigations = useMemo(() => {
    if (!investigations.length) return [];

    const searchTermLower = debouncedSearchTerm?.toLowerCase().trim() || '';
    if (!searchTermLower) return investigations;

    const searchWords = searchTermLower.split(/\s+/);
    const panelSuffix = ` (${t('INVESTIGATION_PANEL')})`.toLowerCase();

    const exactMatch = investigations.find(
      (investigation) =>
        investigation.display.toLowerCase() === searchTermLower ||
        investigation.display.toLowerCase() ===
          `${searchTermLower}${panelSuffix}`,
    );
    if (exactMatch) return [exactMatch];

    return investigations.filter((investigation) => {
      const displayLower = investigation.display.toLowerCase();
      return searchWords.some((word) => displayLower.includes(word));
    });
  }, [investigations, debouncedSearchTerm]);
  return {
    investigations: filteredInvestigations,
    isLoading,
    error,
  };
};

export default useInvestigationsSearch;