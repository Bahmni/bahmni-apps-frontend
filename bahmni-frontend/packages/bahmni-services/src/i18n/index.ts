// Export the default initialized i18n instance
export { default } from './i18n';
