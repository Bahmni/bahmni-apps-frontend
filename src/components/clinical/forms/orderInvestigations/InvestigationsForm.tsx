import React from 'react';
import { ComboBox, Tile } from '@carbon/react';
import { useTranslation } from 'react-i18next';
import { Coding } from 'fhir/r4';
import * as styles from './styles/InvestigationsForm.module.scss';

const InvestigationsForm: React.FC = React.memo(() => {
  const { t } = useTranslation();

  const handleChange = (selectedItem: Coding | null | undefined) => {
    // TODO: Implement investigation selection logic when useInvestigationsSearch hook is ready
    if (selectedItem) {
      // Investigation selected - logic will be implemented with the hook
    }
  };

  return (
    <Tile className={styles.investigationsFormTile}>
      <div className={styles.investigationsFormTitle}>
        {t('INVESTIGATIONS_FORM_TITLE')}
      </div>
      <ComboBox
        id="investigations-search"
        placeholder={t('INVESTIGATIONS_SEARCH_PLACEHOLDER')}
        items={[]}
        onChange={({ selectedItem }) => handleChange(selectedItem)}
        autoAlign
        aria-label={t('INVESTIGATIONS_SEARCH_ARIA_LABEL')}
      />
    </Tile>
  );
});

InvestigationsForm.displayName = 'InvestigationsForm';

export default InvestigationsForm;
