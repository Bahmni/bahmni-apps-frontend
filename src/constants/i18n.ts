export const LOCALE_COOKIE_NAME = "NG_TRANSLATE_LANG_KEY";