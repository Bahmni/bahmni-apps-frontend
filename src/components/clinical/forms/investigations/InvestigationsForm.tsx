import React, { useMemo } from 'react';
import { ComboBox, Tile } from '@carbon/react';
import { useTranslation } from 'react-i18next';
import * as styles from './styles/InvestigationsForm.module.scss';
import useInvestigationsSearch from '@hooks/useInvestigationsSearch';
import type { FlattenedInvestigations } from '@types/investigations';
import useServiceRequestStore from '@stores/serviceRequestStore';
import BoxWHeader from '@components/common/boxWHeader/BoxWHeader';
import SelectedInvestigationItem from './SelectedInvestigationItem';
import SelectedItem from '@components/common/selectedItem/SelectedItem';
import useNotification from '@hooks/useNotification';

const InvestigationsForm: React.FC = React.memo(() => {
  const { t } = useTranslation();
  const [searchTerm, setSearchTerm] = React.useState<string>('');
  const { investigations, isLoading, error } =
    useInvestigationsSearch(searchTerm);
  const {
    selectedServiceRequests,
    addServiceRequest,
    updatePriority,
    removeServiceRequest,
  } = useServiceRequestStore();

  const { addNotification } = useNotification();

  const getFilteredInvestigations = (): FlattenedInvestigations[] => {
    if (searchTerm.length === 0) return [];
    if (isLoading) {
      return [
        {
          code: '',
          display: t('LOADING_CONCEPTS'),
          category: '',
          categoryCode: '',
          disabled: isLoading,
        },
      ];
    }
    if (error) {
      addNotification({
        title: t('ERROR_DEFAULT_TITLE'),
        message: t('ERROR_SEARCHING_INVESTIGATIONS', { error: error.message }),
        type: 'error',
      });
      return [
        {
          code: '',
          display: t('ERROR_SEARCHING_INVESTIGATIONS', {
            error: error.message,
          }),
          category: '',
          categoryCode: '',
          disabled: true,
        },
      ];
    }
    const isSearchEmpty = investigations.length === 0;
    if (isSearchEmpty) {
      return [
        {
          code: '',
          display: t('NO_MATCHING_INVESTIGATIONS_FOUND'),
          category: '',
          categoryCode: '',
          disabled: true,
        },
      ];
    }
    let currentCategory: string | null = null;
    const investigationsCopy = [...investigations];
    for (let i = 0; i < investigationsCopy.length; i++) {
      const investigation = investigationsCopy[i];
      if (investigation.category.toUpperCase() !== currentCategory) {
        currentCategory = investigation.category.toUpperCase();
        investigationsCopy.splice(i, 0, {
          code: '',
          display: currentCategory,
          category: '',
          categoryCode: '',
          disabled: true,
        });
        i++;
      }
    }
    return investigationsCopy;
  };
  const filteredInvestigations: FlattenedInvestigations[] = useMemo(
    () => getFilteredInvestigations(),
    [investigations, searchTerm, isLoading, error],
  );

  const handleChange = (
    selectedItem: FlattenedInvestigations | null | undefined,
  ) => {
    if (selectedItem) {
      addServiceRequest(
        selectedItem.category,
        selectedItem.code,
        selectedItem.display,
      );
    }
  };

  return (
    <Tile className={styles.investigationsFormTile}>
      <div className={styles.investigationsFormTitle}>
        {t('INVESTIGATIONS_FORM_TITLE')}
      </div>
      <ComboBox
        id="investigations-search"
        placeholder={t('INVESTIGATIONS_SEARCH_PLACEHOLDER')}
        items={filteredInvestigations}
        itemToString={(item) => item?.display || ''}
        onChange={({ selectedItem }) => handleChange(selectedItem)}
        onInputChange={(input) => setSearchTerm(input)}
        autoAlign
        aria-label={t('INVESTIGATIONS_SEARCH_ARIA_LABEL')}
      />

      {selectedServiceRequests &&
        selectedServiceRequests.size > 0 &&
        Array.from(selectedServiceRequests.keys()).map((category) => (
          <BoxWHeader
            key={category}
            title={t('INVESTIGATIONS_ADDED', {
              investigationType: category,
            })}
            className={styles.addedInvestigationsBox}
          >
            {selectedServiceRequests.get(category)?.map((serviceRequest) => (
              <SelectedItem
                key={serviceRequest.id}
                onClose={() =>
                  removeServiceRequest(category, serviceRequest.id)
                }
                className={styles.selectedInvestigationItem}
              >
                <SelectedInvestigationItem
                  key={serviceRequest.id}
                  investigation={serviceRequest}
                  onPriorityChange={(priority) =>
                    updatePriority(category, serviceRequest.id, priority)
                  }
                ></SelectedInvestigationItem>
              </SelectedItem>
            ))}
          </BoxWHeader>
        ))}
    </Tile>
  );
});

InvestigationsForm.displayName = 'InvestigationsForm';

export default InvestigationsForm;
