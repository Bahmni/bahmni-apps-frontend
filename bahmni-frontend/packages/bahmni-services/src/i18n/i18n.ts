import i18n from 'i18next';
import LanguageDetector from 'i18next-browser-languagedetector';
import { initReactI18next } from 'react-i18next';
import { getTranslations, getUserPreferredLocale } from './translationService';
import { LOCALE_STORAGE_KEY, CLINICAL_NAMESPACE } from './constants';

/**
 * Initialize i18n with pre-loaded translations
 * This is exported as a function to be called before rendering the app
 */
export const initI18n = async () => {
  const userPreferredLocale = getUserPreferredLocale();
  const translations = await getTranslations(
    userPreferredLocale,
    CLINICAL_NAMESPACE,
  );

  await i18n
    .use(LanguageDetector)
    .use(initReactI18next)
    .init({
      fallbackLng: 'en',
      ns: [CLINICAL_NAMESPACE],
      defaultNS: CLINICAL_NAMESPACE,
      resources: translations,
      detection: {
        order: ['localStorage'],
        lookupLocalStorage: LOCALE_STORAGE_KEY,
      },
      interpolation: {
        escapeValue: false,
      },
      react: {
        useSuspense: true,
      },
    });

  return i18n;
};

// Export the i18n instance for direct access
export const i18nInstance = i18n;

// Export the initialization promise as default
export default await initI18n();
