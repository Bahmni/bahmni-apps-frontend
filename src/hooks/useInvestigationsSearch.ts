import i18next from 'i18next';
import { useEffect, useState, useMemo } from 'react';
import { getFlattenedInvestigations } from '@services/investigationService';
import { FlattenedInvestigations } from '@types/investigations';
import { getFormattedError } from '@utils/common';
import useDebounce from './useDebounce';

interface UseInvestigationsSearchResult {
  investigations: FlattenedInvestigations[];
  isLoading: boolean;
  error: Error | null;
}

const useInvestigationsSearch = (
  searchTerm: string = '',
): UseInvestigationsSearchResult => {
  const [investigations, setInvestigations] = useState<
    FlattenedInvestigations[]
  >([]);
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState<Error | null>(null);
  const debouncedSearchTerm = useDebounce(searchTerm);

  useEffect(() => {
    const fetchInvestigations = async () => {
      try {
        setIsLoading(true);
        setError(null);
        const fetchedInvestigations = await getFlattenedInvestigations();
        setInvestigations(fetchedInvestigations);
      } catch (err) {
        const formattedError = getFormattedError(err);
        setError(new Error(formattedError.message));
      } finally {
        setIsLoading(false);
      }
    };
    fetchInvestigations();
  }, []);

  const filteredInvestigations = useMemo(() => {
    if (!investigations.length) return [];

    const searchTermLower = debouncedSearchTerm?.toLowerCase().trim() || '';
    if (!searchTermLower) return investigations;

    const searchWords = searchTermLower.split(/\s+/);
    const panelSuffix = ` (${i18next.t('INVESTIGATION_PANEL')})`.toLowerCase();

    const exactMatch = investigations.find(
      (investigation) =>
        investigation.display.toLowerCase() === searchTermLower ||
        investigation.display.toLowerCase() ===
          `${searchTermLower}${panelSuffix}`,
    );
    if (exactMatch) return [exactMatch];

    return investigations.filter((investigation) => {
      const displayLower = investigation.display.toLowerCase();
      return searchWords.some((word) => displayLower.includes(word));
    });
  }, [investigations, debouncedSearchTerm]);

  return {
    investigations: filteredInvestigations,
    isLoading,
    error,
  };
};

export default useInvestigationsSearch;
