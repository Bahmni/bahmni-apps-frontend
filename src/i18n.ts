import i18n from 'i18next';
import { initReactI18next } from 'react-i18next';
import Backend from 'i18next-http-backend';
import LanguageDetector from 'i18next-browser-languagedetector';

const mergeTranslations = async (lng: string) => {
  const standardConfigUrl = `bahmni_config/openmrs/i18n/home/locale_${lng}.json`;
  const localConfigUrl = `/locales/locale_${lng}.json`;
  const [standardConfig, localConfig] = await Promise.all([
    fetch(standardConfigUrl).then((res) => res.json()),
    fetch(localConfigUrl).then((res) => res.json()),
  ]);
  return { ...localConfig,  ...standardConfig }; // Local config overrides standard config
};

i18n
  .use(Backend)
  .use(LanguageDetector)
  .use(initReactI18next)
  .init({
    fallbackLng: 'en',
    debug: process.env.NODE_ENV === 'development',
    ns: ['translation'],
    defaultNS: process.env.REACT_APP_DEFAULT_NAMESPACE || 'translation',

    detection: {
      order: ['cookie', 'localStorage', 'navigator', 'htmlTag'],
      lookupCookie: 'NG_TRANSLATE_LANG_KEY', // Match Bahmni's existing cookie name
      caches: ['cookie'],
      cookieOptions: {
        path: '/',
        sameSite: 'strict'
      }},
    interpolation: {
      escapeValue: false,
    },
    backend: {
      loadPath: '/locales/locale_{{lng}}.json',
      parse: async (data: string, lng: string) => {
        const mergedTranslations = await mergeTranslations(lng);
        return mergedTranslations;
      }
    }, // Properly closed backend object
    react: {
      useSuspense: true,
    },
  });

export default i18n;
