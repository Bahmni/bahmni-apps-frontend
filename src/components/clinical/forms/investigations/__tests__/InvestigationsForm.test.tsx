import React from 'react';
import { render, screen, waitFor, act } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { I18nextProvider } from 'react-i18next';
import i18n from '@/setupTests.i18n';
import InvestigationsForm from '../InvestigationsForm';
import { axe, toHaveNoViolations } from 'jest-axe';
import type { FlattenedInvestigations } from '@types/investigations';

expect.extend(toHaveNoViolations);

Element.prototype.scrollIntoView = jest.fn();

jest.mock('../styles/InvestigationsForm.module.scss', () => ({
  investigationsFormTile: 'investigationsFormTile',
  investigationsFormTitle: 'investigationsFormTitle',
  addedInvestigationsBox: 'addedInvestigationsBox',
  selectedInvestigationItem: 'selectedInvestigationItem',
}));

jest.mock('@hooks/useInvestigationsSearch');
jest.mock('@stores/serviceRequestStore');
jest.mock('@hooks/useNotification');
jest.mock('@components/common/boxWHeader/BoxWHeader', () => ({
  __esModule: true,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  default: ({ children, title, className }: any) => (
    <div className={className} data-testid="box-w-header">
      <h3>{title}</h3>
      {children}
    </div>
  ),
}));
jest.mock('@components/common/selectedItem/SelectedItem', () => ({
  __esModule: true,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  default: ({ children, onClose, className }: any) => (
    <div className={className} data-testid="selected-item">
      {children}
      <button onClick={onClose} aria-label="Remove">
        Ã—
      </button>
    </div>
  ),
}));
jest.mock('../SelectedInvestigationItem', () => ({
  __esModule: true,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  default: ({ investigation, onPriorityChange }: any) => (
    <div data-testid="selected-investigation-item">
      <span>{investigation.display}</span>
      <input
        type="checkbox"
        onChange={(e) =>
          onPriorityChange(e.target.checked ? 'stat' : 'routine')
        }
        aria-label="Set as urgent"
      />
    </div>
  ),
}));

import useInvestigationsSearch from '@hooks/useInvestigationsSearch';
import useServiceRequestStore from '@stores/serviceRequestStore';
import useNotification from '@hooks/useNotification';

const mockInvestigations: FlattenedInvestigations[] = [
  {
    code: 'cbc-001',
    display: 'Complete Blood Count',
    category: 'Laboratory',
    categoryCode: 'lab',
  },
  {
    code: 'glucose-001',
    display: 'Blood Glucose Test',
    category: 'Laboratory',
    categoryCode: 'lab',
  },
  {
    code: 'xray-001',
    display: 'Chest X-Ray',
    category: 'Radiology',
    categoryCode: 'rad',
  },
];
const mockStore = {
  selectedServiceRequests: new Map(),
  addServiceRequest: jest.fn(),
  removeServiceRequest: jest.fn(),
  updatePriority: jest.fn(),
  reset: jest.fn(),
  getState: jest.fn(() => ({
    selectedServiceRequests: new Map(),
  })),
};

describe('InvestigationsForm', () => {
  const mockAddNotification = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
    // Setup default mocks
    (useInvestigationsSearch as jest.Mock).mockReturnValue({
      investigations: [],
      isLoading: false,
      error: null,
    });

    (useServiceRequestStore as unknown as jest.Mock).mockReturnValue(mockStore);

    (useNotification as jest.Mock).mockReturnValue({
      addNotification: mockAddNotification,
    });
  });

  describe('Component Rendering', () => {
    test('renders form with title and search combobox', () => {
      render(
        <I18nextProvider i18n={i18n}>
          <InvestigationsForm />
        </I18nextProvider>,
      );

      expect(screen.getByText('Order Investigations')).toBeInTheDocument();
      const combobox = screen.getByRole('combobox');
      expect(combobox).toBeInTheDocument();
      expect(combobox).toHaveAttribute('id', 'investigations-search');
      expect(combobox).toHaveAttribute(
        'placeholder',
        'Search to add laboratory or radiology investigations',
      );
      expect(combobox).toHaveAttribute(
        'aria-label',
        'INVESTIGATIONS_SEARCH_ARIA_LABEL',
      );
    });
  });

  describe('Search Functionality', () => {
    test('updates search term on input change', async () => {
      const user = userEvent.setup();

      render(
        <I18nextProvider i18n={i18n}>
          <InvestigationsForm />
        </I18nextProvider>,
      );

      const combobox = screen.getByRole('combobox');
      await user.type(combobox, 'blood');

      expect(combobox).toHaveValue('blood');
    });

    test('displays loading state when searching', async () => {
      (useInvestigationsSearch as jest.Mock).mockReturnValue({
        investigations: [],
        isLoading: true,
        error: null,
      });

      const user = userEvent.setup();

      render(
        <I18nextProvider i18n={i18n}>
          <InvestigationsForm />
        </I18nextProvider>,
      );

      const combobox = screen.getByRole('combobox');
      await user.type(combobox, 'test');
      await waitFor(() => {
        expect(screen.getByText(/loading concepts.../i)).toBeInTheDocument();
      });
    });

    test('displays error state when search fails', async () => {
      const mockError = new Error('Failed to fetch investigations');
      (useInvestigationsSearch as jest.Mock).mockReturnValue({
        investigations: [],
        isLoading: false,
        error: mockError,
      });

      const user = userEvent.setup();

      render(
        <I18nextProvider i18n={i18n}>
          <InvestigationsForm />
        </I18nextProvider>,
      );

      const combobox = screen.getByRole('combobox');
      await user.type(combobox, 'test');

      expect(mockAddNotification).toHaveBeenCalledWith({
        title: 'Error',
        message: 'Error searching investigations',
        type: 'error',
      });
      await waitFor(() => {
        expect(
          screen.getByText(/error searching investigations/i),
        ).toBeInTheDocument();
      });
    });

    test('displays no results message when search returns empty', async () => {
      (useInvestigationsSearch as jest.Mock).mockReturnValue({
        investigations: [],
        isLoading: false,
        error: null,
      });

      const user = userEvent.setup();

      render(
        <I18nextProvider i18n={i18n}>
          <InvestigationsForm />
        </I18nextProvider>,
      );

      const combobox = screen.getByRole('combobox');
      await user.type(combobox, 'nonexistent');
      await waitFor(() => {
        expect(
          screen.getByText(/no matching investigations found/i),
        ).toBeInTheDocument();
      });
    });

    test('displays investigations grouped by category', async () => {
      (useInvestigationsSearch as jest.Mock).mockReturnValue({
        investigations: mockInvestigations,
        isLoading: false,
        error: null,
      });

      const user = userEvent.setup();

      render(
        <I18nextProvider i18n={i18n}>
          <InvestigationsForm />
        </I18nextProvider>,
      );

      const combobox = screen.getByRole('combobox');
      await user.type(combobox, 'test');

      // The investigations are filtered and grouped in the component
      expect(combobox).toHaveValue('test');
      await waitFor(() => {
        const options = screen.getAllByRole('option');
        expect(options).toHaveLength(5);
        expect(options[0]).toHaveTextContent('LABORATORY');
        expect(options[1]).toHaveTextContent('Complete Blood Count');
        expect(options[2]).toHaveTextContent('Blood Glucose Test');
        expect(options[3]).toHaveTextContent('RADIOLOGY');
        expect(options[4]).toHaveTextContent('Chest X-Ray');
      });
    });
  });

  describe('Investigation Selection', () => {
    test('adds investigation when selected from dropdown', async () => {
      const user = userEvent.setup();
      (useInvestigationsSearch as jest.Mock).mockReturnValue({
        investigations: mockInvestigations,
        isLoading: false,
        error: null,
      });

      render(
        <I18nextProvider i18n={i18n}>
          <InvestigationsForm />
        </I18nextProvider>,
      );

      // Simulate selecting an investigation by calling the onChange handler
      const combobox = screen.getByRole('combobox');

      await waitFor(async () => {
        await user.type(combobox, 'complete');
      });

      // Wait for the dropdown item to appear
      await waitFor(() => {
        expect(screen.getByText('Complete Blood Count')).toBeInTheDocument();
      });

      // Click on the dropdown item
      await waitFor(async () => {
        await user.click(screen.getByText('Complete Blood Count'));
      });

      // Verify the store was called correctly
      await waitFor(() => {
        expect(mockStore.addServiceRequest).toHaveBeenCalledWith(
          'Laboratory',
          'cbc-001',
          'Complete Blood Count',
        );
      });
    });
  });

  describe('Selected Investigations Display', () => {
    test('displays selected investigations grouped by category', () => {
      const selectedMap = new Map([
        [
          'Laboratory',
          [
            {
              id: 'cbc-001',
              display: 'Complete Blood Count',
              selectedPriority: 'routine',
            },
            {
              id: 'glucose-001',
              display: 'Blood Glucose Test',
              selectedPriority: 'stat',
            },
          ],
        ],
        [
          'Radiology',
          [
            {
              id: 'xray-001',
              display: 'Chest X-Ray',
              selectedPriority: 'routine',
            },
          ],
        ],
      ]);

      (useServiceRequestStore as unknown as jest.Mock).mockReturnValue({
        ...mockStore,
        selectedServiceRequests: selectedMap,
      });

      render(
        <I18nextProvider i18n={i18n}>
          <InvestigationsForm />
        </I18nextProvider>,
      );

      // Check category headers
      expect(screen.getByText('Added Laboratory')).toBeInTheDocument();
      expect(screen.getByText('Added Radiology')).toBeInTheDocument();

      // Check investigations
      expect(screen.getByText('Complete Blood Count')).toBeInTheDocument();
      expect(screen.getByText('Blood Glucose Test')).toBeInTheDocument();
      expect(screen.getByText('Chest X-Ray')).toBeInTheDocument();
    });

    test('removes investigation when close button is clicked', async () => {
      const selectedMap = new Map([
        [
          'Laboratory',
          [
            {
              id: 'cbc-001',
              display: 'Complete Blood Count',
              selectedPriority: 'routine',
            },
          ],
        ],
      ]);

      (useServiceRequestStore as unknown as jest.Mock).mockReturnValue({
        ...mockStore,
        selectedServiceRequests: selectedMap,
      });

      const user = userEvent.setup();

      render(
        <I18nextProvider i18n={i18n}>
          <InvestigationsForm />
        </I18nextProvider>,
      );

      const removeButton = screen.getByLabelText('Remove');
      await user.click(removeButton);

      expect(mockStore.removeServiceRequest).toHaveBeenCalledWith(
        'Laboratory',
        'cbc-001',
      );
    });

    test('updates priority when urgent checkbox is toggled', async () => {
      const selectedMap = new Map([
        [
          'Laboratory',
          [
            {
              id: 'cbc-001',
              display: 'Complete Blood Count',
              selectedPriority: 'routine',
            },
          ],
        ],
      ]);

      (useServiceRequestStore as unknown as jest.Mock).mockReturnValue({
        ...mockStore,
        selectedServiceRequests: selectedMap,
      });

      const user = userEvent.setup();

      render(
        <I18nextProvider i18n={i18n}>
          <InvestigationsForm />
        </I18nextProvider>,
      );

      const urgentCheckbox = screen.getByLabelText('Set as urgent');
      await user.click(urgentCheckbox);

      expect(mockStore.updatePriority).toHaveBeenCalledWith(
        'Laboratory',
        'cbc-001',
        'stat',
      );
    });
  });

  describe('Edge Cases', () => {
    test('handles empty search term correctly', () => {
      render(
        <I18nextProvider i18n={i18n}>
          <InvestigationsForm />
        </I18nextProvider>,
      );

      const combobox = screen.getByRole('combobox');
      expect(combobox).toHaveValue('');
    });
    test('should handle search result with empty display', async () => {
      (useInvestigationsSearch as jest.Mock).mockReturnValue({
        investigations: [
          {
            code: 'empty-001',
            display: '',
            category: 'Laboratory',
            categoryCode: 'lab',
          },
        ],
        isLoading: false,
        error: null,
      });

      render(
        <I18nextProvider i18n={i18n}>
          <InvestigationsForm />
        </I18nextProvider>,
      );

      const combobox = screen.getByRole('combobox');
      const user = userEvent.setup();
      await user.type(combobox, 'test');
      expect(combobox).toHaveValue('test');
      expect(screen.getByRole('option', { name: '' })).toBeInTheDocument();
      expect(
        screen.getByRole('option', { name: 'LABORATORY' }),
      ).toBeInTheDocument();
    });
  });

  describe('Accessibility', () => {
    test('has no accessibility violations', async () => {
      let container: HTMLElement;

      await act(async () => {
        const rendered = render(
          <I18nextProvider i18n={i18n}>
            <InvestigationsForm />
          </I18nextProvider>,
        );
        container = rendered.container;
      });

      const results = await axe(container!);
      expect(results).toHaveNoViolations();
    });
  });
});
