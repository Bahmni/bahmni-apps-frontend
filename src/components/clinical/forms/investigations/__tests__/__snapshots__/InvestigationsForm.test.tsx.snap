// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Tests matches snapshot with default state 1`] = `
<div>
  <div
    class="cds--tile investigationsFormTile"
  >
    <div
      class="investigationsFormTitle"
    >
      Order Investigations
    </div>
    <div
      class="cds--list-box__wrapper"
    >
      <div
        class="cds--combo-box cds--autoalign cds--list-box"
      >
        <div
          class="cds--list-box__field"
        >
          <input
            aria-activedescendant=""
            aria-autocomplete="list"
            aria-controls="downshift-«r1c»-menu"
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-label="INVESTIGATIONS_SEARCH_ARIA_LABEL"
            autocomplete="off"
            class="cds--text-input cds--text-input--empty"
            id="investigations-search"
            placeholder="Search to add laboratory or radiology investigations"
            role="combobox"
            tabindex="0"
            title=""
            type="text"
            value=""
          />
          <button
            aria-controls="downshift-«r1c»-menu"
            aria-expanded="false"
            aria-label="Open"
            class="cds--list-box__menu-icon"
            id="downshift-«r1c»-toggle-button"
            tabindex="-1"
            title="Open"
            type="button"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              focusable="false"
              height="16"
              preserveAspectRatio="xMidYMid meet"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"
              />
            </svg>
          </button>
        </div>
        <ul
          aria-labelledby="downshift-«r1c»-label"
          class="cds--list-box__menu"
          id="downshift-«r1c»-menu"
          role="listbox"
          style="position: fixed; left: 0px; top: 0px; visibility: visible; transform: translate(0px, 0px);"
        />
      </div>
    </div>
  </div>
</div>
`;
